pkg/osquery/table/touchid_user_darwin.go

```diff
@@ -0,0 +1,143 @@
+package table
+
+import (
+	"bytes"
+	"context"
+	"os/exec"
+	"os/user"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/kolide/osquery-go"
+	"github.com/kolide/osquery-go/plugin/table"
+)
+
+func TouchIDUserConfig(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+	t := &touchIDUserConfigTable{
+		client: client,
+		logger: logger,
+	}
+	columns := []table.ColumnDefinition{
+		table.IntegerColumn("uid"),
+		table.IntegerColumn("fingerprints_registered"),
+		table.IntegerColumn("touchid_unlock"),
+		table.IntegerColumn("touchid_applepay"),
+		table.IntegerColumn("effective_unlock"),
+		table.IntegerColumn("effective_applepay"),
+	}
+
+	return table.NewPlugin("kolide_touchid_user_config", columns, t.generate)
+}
+
+type touchIDUserConfigTable struct {
+	client *osquery.ExtensionManagerClient
+	logger log.Logger
+	config *touchIDUserConfig
+}
+
+type touchIDUserConfig struct {
+	uid                    int
+	fingerprintsRegistered int
+	touchIDUnlock          int
+	touchIDApplePay        int
+	effectiveUnlock        int
+	effectiveApplePay      int
+}
+
+func (t *touchIDUserConfigTable) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	q, _ := queryContext.Constraints["uid"]
+	if len(q.Constraints) == 0 {
+		level.Debug(t.logger).Log(
+			"msg", "The touchid_user_config table requires that you specify a constraint WHERE uid =",
+			"err", "no constraints",
+		)
+		return nil, errors.New("The touchid_user_config table requires that you specify a constraint WHERE uid =")
+	}
+
+	var results []map[string]string
+	for _, constraint := range q.Constraints {
+		var touchIDUnlock, touchIDApplePay, effectiveUnlock, effectiveApplePay string
+		var stdout bytes.Buffer
+
+		// Verify the user exists on the system before proceeding
+		_, err := user.LookupId(constraint.Expression)
+		if err != nil {
+			level.Debug(t.logger).Log(
+				"msg", "nonexistant user",
+				"uid", constraint.Expression,
+				"err", err,
+			)
+			continue
+		}
+		uid, _ := strconv.Atoi(constraint.Expression)
+
+		// Grab the user's TouchID configuration
+		cmd := exec.Command("/usr/bin/bioutil", "-r")
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+		cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(uid), Gid: 20}
+		cmd.Stdout = &stdout
+		if err := cmd.Run(); err != nil {
+			level.Debug(t.logger).Log(
+				"msg", "Failed to run bioutil for configuration",
+				"uid", uid,
+				"err", err,
+			)
+			continue
+		}
+		configOutStr := string(stdout.Bytes())
+		configSplit := strings.Split(configOutStr, ":")
```

Comment:
The exec and parse `bioutil` is repeated. What do you think about pulling into a shared function with an arg for system or user? (not a full package)

Reply:
I don't think that is a good idea, at least for now. To my knowledge, with the current way of parsing the output, it would just move the code into a different function and maybe save a few lines of code. I think the way it is laid out currently is easier to read even if slightly less efficient.

Reply:
I'm thinking of the regext parsing code, and positional parameters. I'm pretty sure if I find a bug in 6 months, I'll forget to change one of them